diff --git a/Makefile b/Makefile
index 921568f2187d0fc2eb1218a0c56609f70ce82a9e..058e1c2a8339bf1fc50d3b6690717d474675951b 100644
--- a/Makefile
+++ b/Makefile
@@ -24,7 +24,7 @@
 VERSION = 2013
 PATCHLEVEL = 04
 SUBLEVEL =
-EXTRAVERSION = -sphairon4.17
+EXTRAVERSION = -sphairon4.18
 ifneq "$(SUBLEVEL)" ""
 U_BOOT_VERSION = $(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)
 else
diff --git a/board/sphairon/isdn_sip_gw/isdn_sip_gw.c b/board/sphairon/isdn_sip_gw/isdn_sip_gw.c
index aa78483c1c3d2917b7013f55bac491a182ec7f65..99267f270d145127e73607c15dd4344de07d7d09 100644
--- a/board/sphairon/isdn_sip_gw/isdn_sip_gw.c
+++ b/board/sphairon/isdn_sip_gw/isdn_sip_gw.c
@@ -82,8 +82,8 @@ static inline void gpio_init(void)
 	/* PCM DCL/FSC input/output switch -> default output */
 	gpio_direction_output(47, 1);
 
-        /* High output for supervisor */
-        gpio_direction_output(8, 1);
+	/* High output for supervisor */
+	gpio_direction_output(8, 1);
 }
 
 int board_early_init_f(void)
@@ -247,10 +247,10 @@ void sas_cf_led_action(enum sas_cf_state state)
 	return;
 }
 
-int sas_board_init_r(void)
+int sas_board_late_init_r(void)
 {
 	/* enable CLKOUT0 for Supervisor */
-        gpio_set_altfunc(8, GPIO_ALTSEL_SET, GPIO_ALTSEL_CLR, GPIO_DIR_OUT);
+	gpio_set_altfunc(8, GPIO_ALTSEL_SET, GPIO_ALTSEL_CLR, GPIO_DIR_OUT);
 
 	return 0;
 }
